### ISSUE TITLE
Subsequent correction iterations lose information about cached offense status

### ISSUE THREAD

**Issue by viralpraxis** [rubocop/rubocop#14222]
I don't think this is a big deal, but in case someone else gets confused by this behavior they'll be able to find this issue (feel free to close if it was already discussed).

## Steps to reproduce the problem

```shell
> rm -rf ~/.cache/rubocop_cache
> bundle exec rubocop          
Inspecting 27 files
........C..................

Offenses:

lib/rubocop/cop/thread_safety/class_instance_variable.rb:230:60: C: [Correctable] InternalAffairs/NodePatternGroups: Replace def, defs in node pattern union with any_def.
            (block (send (const nil? :Struct) :new ...) _ ({def defs} ...))
                                                           ^^^^^^^^^^
lib/rubocop/cop/thread_safety/class_instance_variable.rb:231:59: C: InternalAffairs/NodePatternGroups: Replace def, defs in node pattern union with any_def.
            (block (send (const nil? :Class) :new ...) _ ({def defs} ...))
                                                          ^^^^^^^^^^
lib/rubocop/cop/thread_safety/class_instance_variable.rb:232:61: C: InternalAffairs/NodePatternGroups: Replace def, defs in node pattern union with any_def.
            (block (send (const nil? :Data) :define ...) _ ({def defs} ...))
```

Note that only the first offense is marked as `[correctable]` -- this is [expected](https://github.com/rubocop/rubocop/blob/2e91abd41e126f6c9d82efdf72cedb044285cc41/lib/rubocop/cop/internal_affairs/node_pattern_groups.rb#L99-L102)

Now run `bundle exec rubocop -A --cache true` to autocorrect all of these issues.

Now restore these autocorrections (e.g. with `git restore lib`) and run RuboCop with and without cache:

```shell
> bundle exec rubocop --cache true             
Inspecting 27 files
........C..................

Offenses:

lib/rubocop/cop/thread_safety/class_instance_variable.rb:230:60: C: [Correctable] InternalAffairs/NodePatternGroups: Replace def, defs in node pattern union with any_def.
            (block (send (const nil? :Struct) :new ...) _ ({def defs} ...))
                                                           ^^^^^^^^^^
lib/rubocop/cop/thread_safety/class_instance_variable.rb:231:56: C: [Correctable] InternalAffairs/NodePatternGroups: Replace def, defs in node pattern union with any_def.
            (block (send (const nil? :Class) :new ...) _ ({def defs} ...))
                                                       ^^^^^^^^^^
lib/rubocop/cop/thread_safety/class_instance_variable.rb:232:55: C: [Correctable] InternalAffairs/NodePatternGroups: Replace def, defs in node pattern union with any_def.
            (block (send (const nil? :Data) :define ...) _ ({def defs} ...))
                                                      ^^^^^^^^^^

27 files inspected, 3 offenses detected, 3 offenses autocorrectable
>  bundle exec rubocop --cache false
Inspecting 27 files
........C..................

Offenses:

lib/rubocop/cop/thread_safety/class_instance_variable.rb:230:60: C: [Correctable] InternalAffairs/NodePatternGroups: Replace def, defs in node pattern union with any_def.
            (block (send (const nil? :Struct) :new ...) _ ({def defs} ...))
                                                           ^^^^^^^^^^
lib/rubocop/cop/thread_safety/class_instance_variable.rb:231:59: C: InternalAffairs/NodePatternGroups: Replace def, defs in node pattern union with any_def.
            (block (send (const nil? :Class) :new ...) _ ({def defs} ...))
                                                          ^^^^^^^^^^
lib/rubocop/cop/thread_safety/class_instance_variable.rb:232:61: C: InternalAffairs/NodePatternGroups: Replace def, defs in node pattern union with any_def.
            (block (send (const nil? :Data) :define ...) _ ({def defs} ...))
                                                            ^^^^^^^^^^

27 files inspected, 3 offenses detected, 1 offense autocorrectable
```

Running with cache shows all the offenses as autocorrectable, and running without caching shows only the first one.

## RuboCop version

latest commit

**Comment by viralpraxis** [comment]
Also, it seems like it ruins ^highlight a little bit:

![Image](https://github.com/user-attachments/assets/6f3a8adf-ab4c-4a78-a07c-2c9be69e8ed6)

**Comment by bbatsov** [comment]
I can't reproduce this on current master - the result cache now preserves each offense's correctable status and column ranges, so cached and uncached runs come out identical. Looks like it was fixed somewhere along the way. Closing, but please reopen if you still hit it on the latest version.
